logic/odmr_logic.py

```diff
@@ -454,41 +496,69 @@ def mw_sweep_on(self):
 
         limits = self.get_hw_constraints()
         param_dict = {}
-
+        self.frequency_lists = []
+        self.final_freq_list = []
         if self.mw_scanmode == MicrowaveMode.LIST:
-            if np.abs(self.mw_stop - self.mw_start) / self.mw_step >= limits.list_maxentries:
-                self.log.warning('Number of frequency steps too large for microwave device. '
-                                 'Lowering resolution to fit the maximum length.')
-                self.mw_step = np.abs(self.mw_stop - self.mw_start) / (limits.list_maxentries - 1)
-                self.sigParameterUpdated.emit({'mw_step': self.mw_step})
-
-            # adjust the end frequency in order to have an integer multiple of step size
-            # The master module (i.e. GUI) will be notified about the changed end frequency
-            num_steps = int(np.rint((self.mw_stop - self.mw_start) / self.mw_step))
-            end_freq = self.mw_start + num_steps * self.mw_step
-            freq_list = np.linspace(self.mw_start, end_freq, num_steps + 1)
-            freq_list, self.sweep_mw_power, mode = self._mw_device.set_list(freq_list,
-                                                                            self.sweep_mw_power)
-            self.mw_start = freq_list[0]
-            self.mw_stop = freq_list[-1]
-            self.mw_step = (self.mw_stop - self.mw_start) / (len(freq_list) - 1)
+            final_freq_list = []
+            used_starts = []
+            used_steps = []
+            used_stops = []
+            for mw_start, mw_stop, mw_step in zip(self.mw_starts, self.mw_stops, self.mw_steps):
 
-            param_dict = {'mw_start': self.mw_start, 'mw_stop': self.mw_stop,
-                          'mw_step': self.mw_step, 'sweep_mw_power': self.sweep_mw_power}
+                num_steps = int(np.rint((mw_stop - mw_start) / mw_step))
+                end_freq = mw_start + num_steps * mw_step
+                freq_list = np.linspace(mw_start, end_freq, num_steps + 1)
+                if np.abs(mw_stop - mw_start) / mw_step >= limits.list_maxentries:
+                    self.log.warning('Number of frequency steps too large for microwave device. '
+                                     'Lowering resolution to fit the maximum length.')
+                    mw_step = np.abs(mw_stop - mw_start) / (limits.list_maxentries - 1)
```

Comment:
This is using the length of each range individually but the total length of all ranges should be checked here. So check the total length somewhere and then either reduce the resolution or raise an error.